certora/specs/GeneralAdapter1Reverts.spec

```
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+using WETH9 as WETH;
+using Morpho as Morpho;
+using ERC20Mock as ERC20Mock;
+using ERC20USDT as ERC20USDT;
+using ERC20NoRevert as ERC20NoRevert;
+using Bundler3 as Bundler3;
+using AllowanceTransfer as AllowanceTransfer;
+
+methods {
+    function _.depositFor(address, uint256) external => DISPATCHER(true);
+    function _.withdrawTo(address, uint256) external => DISPATCHER(true);
+    function _.mint(uint256, address) external  => DISPATCHER(true);
+    function _.deposit(uint256, address) external => DISPATCHER(true);
+    function _.withdraw(uint256, address, address) external => DISPATCHER(true);
+    function _.redeem(uint256, address, address) external => DISPATCHER(true);
+    function _.approve(address, uint256)  external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function AllowanceTransfer.allowance(address, address, address) external returns (uint160, uint48, uint48) envfree;
+    function Bundler3.initiator() external returns address envfree;
+}
+
+// Permit2Lib.PERMIT2 concrete address
+definition PERMIT2() returns address = 0x000000000022D473030F116dDEE9F6B43aC78BA3;
+
+// Check that the state didn't change upon a native transfer with 0 ETH using the adapter.
+rule nativeTransferChange(env e, address receiver, uint256 amount) {
+    storage storageBefore = lastStorage;
+    uint256 senderBalanceBefore = nativeBalances[currentContract];
+
+    nativeTransfer(e, receiver, amount);
+
+    assert amount == max_uint256 && senderBalanceBefore == 0 => storageBefore == lastStorage;
```

```suggestion
    assert amount == max_uint256 && senderBalanceBefore == 0 <=> storageBefore == lastStorage;
```
let's prove the equivalence (proves that it's the only case)

Otherwise, we can do this (which is already better than current)
```suggestion
    senderBalanceBefore != 0 => storageBefore != lastStorage;
```

---

About
```
senderBalanceBefore != 0 => storageBefore != lastStorage;
```
Generally, the spirit of these rules (as pointed out by @QGarchery) is not to prove correction but rather to show why they don't hold in the general rule. I'm not sure it's worth checking the equivalences everywhere here. Alternatively, the previous PR was doing a better job at this as it proved the correctness (i.e. the expected state after executing the function). Also, equivalences in Certora are a bit tricky, writing `<=>` is the same as `==`, it doesn't actually check one way and then the other but rather if both expressions evaluate to the same value ([Certora doc](https://docs.certora.com/en/latest/docs/cvl/expr.html#extended-logical-operations)).

---

>why they don't hold in the general rule

the problem that i'm pointing is that here you show one of the reason why they don't hold, not all of them (or at least it's not proven that it's all of the reasons)

>Also, equivalences in Certora are a bit tricky, ...

what does it change, concretely?

---

> the problem that i'm pointing is that here you show one of the reason why they don't hold, not all of them (or at least it's not proven that it's all of the reasons)

this is much more clear to me. The change is done in https://github.com/morpho-org/bundler3/pull/229/commits/57487eb87f39040bf6ed47554a7c9a926a6a4ea3.

---

Done in https://github.com/morpho-org/bundler3/pull/229/commits/fac59f9bf3a36db94f5f563950517bfad8fb410e and https://github.com/morpho-org/bundler3/pull/229/commits/3e16377933769262288e8baab2e74bc42e0aae4e.

> the problem that i'm pointing is that here you show one of the reason why they don't hold, not all of them (or at least it's not proven that it's all of the reasons)

This makes sense, I didn't put it in first hand as there are some corner cases you need to express formally (for instance in `permi2TransferFrom` you need to assume that the `initiator` is not the Permit2 implementation contract, aka `AllowanceTransfer`) but that make no sense in the real world. The point was only to provide a counter-example to explain why it doesn't fit into the general rule. 

> what does it change, concretely?

What I had in mind is that there is something counter intuitive. In the CVL expression `a <=> b`, if both `b` and `a` are always `false` then the result is `true`, which is not "constructive" in the sense, that `a` and `b` needn't be `true` at some point for the property to hold.